File: src/main/java/com/gmail/borlandlp/minigamesdtools/activepoints/type/block/StaticBlockPoint.java
package com.gmail.borlandlp.minigamesdtools.activepoints.type.block;

import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.activepoints.ActivePoint;
import com.gmail.borlandlp.minigamesdtools.activepoints.BuildSchema;
import org.bukkit.Location;
import org.bukkit.Material;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public abstract class StaticBlockPoint extends ActivePoint {
    private List<Location> usedBlocks = new ArrayList<>();
    private String direction;
    private int radius;
    private double health;

    public double getHealth() {
        return health;
    }

    public void setHealth(double health) {
        this.health = health;
    }

    public int getRadius() {
        return radius;
    }

    public void setRadius(int radius) {
        this.radius = radius;
    }

    public void setDirection(String direction) {
        this.direction = direction;
    }

    public String getDirection() {
        return direction;
    }

    public void setUsedBlocks(List<Location> usedBlocks) {
        this.usedBlocks = usedBlocks;
    }

    public List<Location> getUsedBlocks() {
        return usedBlocks;
    }

    public void spawn() {
        Map<Location, Material> buildSchema = this.getBuildSchema().getSchema();
        for(Location location : buildSchema.keySet()) {
            try {
                location.getWorld().getBlockAt(location).setType(buildSchema.get(location));
            } catch (Exception e) {
                e.printStackTrace();
            }
            this.getActivePointController().getStaticPointsCache().add(location, this);
        }
        Debug.print(Debug.LEVEL.NOTICE, "spawn static point " + this.getName() + ". size:" + buildSchema.keySet().size());
        this.setUsedBlocks(new ArrayList<>(buildSchema.keySet()));
        this.setSpawned(true);
    }

    public void despawn() {
        List<Location> buildSchema = this.getUsedBlocks();
        for(Location location : buildSchema) {
            location.getWorld().getBlockAt(location).setType(Material.AIR);
            this.getActivePointController().getStaticPointsCache().remove(location);
        }
        Debug.print(Debug.LEVEL.NOTICE, "despawn static point " + this.getName() + ". size:" + this.getUsedBlocks().size());
        this.setSpawned(false);
    }

    public abstract BuildSchema getBuildSchema();
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/activepoints/type/block/StaticBlockPoint.kt
package com.gmail.borlandlp.minigamesdtools.activepoints.type.block

import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.Debug.print
import com.gmail.borlandlp.minigamesdtools.activepoints.ActivePoint
import com.gmail.borlandlp.minigamesdtools.activepoints.BuildSchema
import org.bukkit.Location
import org.bukkit.Material
import java.util.*

abstract class StaticBlockPoint : ActivePoint() {
    var usedBlocks: List<Location> = ArrayList()
    var direction: String? = null
    var radius = 0
    override var health = 0.0

    override fun spawn() {
        val buildSchema = buildSchema.schema
        for (location in buildSchema.keys) {
            try {
                location.world.getBlockAt(location).type = buildSchema[location]
            } catch (e: Exception) {
                e.printStackTrace()
                print(Debug.LEVEL.WARNING, "${javaClass::getSimpleName}->spawn() :: arena:${activePointController} :: Missing block in BuildSchema for $location")
            }
            activePointController!!.staticPointsCache.add(location, this)
        }
        print(Debug.LEVEL.NOTICE, "spawn static point $name. size:${buildSchema.keys.size}")
        usedBlocks = ArrayList(buildSchema.keys)
        isSpawned = true
    }

    override fun despawn() {
        usedBlocks.apply {
            location!!.world.getBlockAt(location).type = Material.AIR
            activePointController!!.staticPointsCache.remove(location!!)
        }.also {
            print(Debug.LEVEL.NOTICE, "despawn static point $name. size:${usedBlocks.size}")
        }

        isSpawned = false
    }

    abstract val buildSchema: BuildSchema
}